### PR TITLE
Modify kubevirt_vm crud/wait logic

### PR DESCRIPTION
**Pull request by mmazur**
##### SUMMARY
1. Adds proper wait support for VM stops and starts
2. Detect https://github.com/kubevirt/ansible-kubevirt-modules/issues/177 and return a sane error
3. Switch to openshift-restclient 0.9.x style wait code

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
kubevirt/kubevirt_vm

### CODE REVIEW

**Comment by mmazur** [comment]
@machacekondra @pkliczewski – our e2e testing works ok: https://travis-ci.org/kubevirt/ansible-kubevirt-modules/jobs/511551711

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py [hunk 1]
```diff
@@ -249,92 +256,99 @@ def argspec(self):
...
+                definition['spec']['running'] = state == 'running'
+            template = definition['spec']['template']
+            kind = 'VirtualMachineInstance' if ephemeral else 'VirtualMachine'
+            dummy, definition = self.construct_vm_definition(kind, definition, template)
+            result = self.execute_crud(kind, definition)
+            changed = result['changed']
+            k8s_obj = result['result']
+            method = result['method']
+
+        if not ephemeral and state in ['running', 'stopped'] and not self.check_mode:
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py]: So we don't wait for empheral VM? I think we should

Comment (machacekondra) [reply lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py]: Also we should still call this in case of check mode, and say whether it will be changed

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py [hunk 2]
```diff
@@ -249,92 +256,99 @@ def argspec(self):
...
-        # Manage state of the VM:
-        if state in ['running', 'stopped']:
-            if not self.check_mode:
-                ret = self.manage_state(state)
-                changed = changed or ret
+        if ephemeral and state not in ('present', 'absent'):
+            raise TypeError("Ephemeral VMs only support 'present' and 'absent' states.")
+
+        _used_params = [name for name in self.params if self.params[name] is not None]
+        vm_spec_change = True if set(VM_SPEC_PARAMS).intersection(_used_params) else False
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py:357]: What about inline stuff?

Comment (mmazur) [reply lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py:357]: You mean `resource_definition`? That's contained within VM_SPEC_PARAMS.

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py (hunk 1 above)
Comment (mmazur) [reply lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py]: 1. Previous logic didn't wait for ephemeral VMIs either, so no functionality has been lost in that regard.
2. I'm not sure it's possible to wait for VMI shutdowns (with the current openshift-restclient lib), and having only partial support for ephemeral waits is not necessarily better than just having no support.
3. manage_state() call was also skipped when in check_mode, so this PR doesn't change anything there either.

In other words: while your suggestions can be attempted/considered, they are out of scope for this PR, as it was not my goal to add them here and they weren't present before, so no functionality is lost.

Comment (machacekondra) [reply lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py]: 1. It did. When I was last testing.
2. What's missing there? Just waiting for missing entity is ok, no?
3. That's right. But why not support it?

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py (hunk 2 above)
Comment (machacekondra) [reply lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py:357]: Yeah, but does this code works well, when updating some attribute via that? As far as I read it correctly it doesn't.

Comment (mmazur) [reply lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py:357]: It does. If you call the module with any parameter that *might* modify spec: (and that includes `resource_definition`, which doesn't have to modify spec:, but it *might*), that's enough to trigger a full crud call. Only if the module is certain that none of the arguments used touch spec: in any way, does it skip the crud call.

I just pushed a new version rebased on top of master, which includes the module params cleanup PR and that has a much simpler `VM_SPEC_PARAMS` definition, so maybe that'll be clearer.

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py (hunk 1 above)
Comment (mmazur) [reply lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py]: 1. In your code, `manage_state()` is called only `if state in ['running', 'stopped']:`, which aren't really valid/all states for VMIs. And the one thing I tried, failed[1]. So if there are scenarios where your wait code is working on VMIs, I was not able to find them. As far as I could tell, it never did work, which is why I didn't reimplement it in my code.
2. Yes, I'm not aware of how to do waiting for missing entity. The pvc module doesn't have that either. If a way was found, that would be a good thing to add both here for VMIs and in pvc.
3. I'm not saying we shouldn't, I'm just saying I'm not going to in this particular PR, as I have other stuff with higher priority and it wasn't my goal to implement that particular feature here.


[1] Current devel branch, vmi is Running:
```
[mmazur@klapek playbook]$ tail -n 3 vm-start.yml
        wait: yes
        ephemeral: yes
        state: running
[mmazur@klapek playbook]$ apb vm-start.yml
(…)
TASK [start] ************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "error": 422, "msg": "Failed to patch object: b'{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"admission webhook \\\\\"virtualmachineinstances-update-validator.kubevirt.io\\\\\" denied the request: spec.template in body is a forbidden property\",\"reason\":\"Invalid\",\"details\":{\"causes\":[{\"message\":\"spec.template in body is a forbidden property\"}]},\"code\":422}\\n'", "reason": "Unprocessable Entity", "status": 422}
```

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py (hunk 2 above)
Comment (machacekondra) [reply lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py:357]: I see now. Quite complex.. If we want to say user he can't stop/start non-existing VM, wouldn't it be simpler to just say that(one simple condition) and instead of saying that he sent empty parameters tell him, which parameters are required(returned from backend) this would very simplify all of this and I don't think we really need this. But if you insist this is better I am ok.

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py (hunk 1 above)
Comment (machacekondra) [reply lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py]: 1. That's issue of creating whole ephemral VM: https://github.com/ansible/ansible/pull/54453 With this patch empheral VMs+waiting works ok again. It was broken by template module, where I had to rewrote most of the things, so it worked, but there was just a bug.

2. Waiting for 404 could work. I used that in other modules.

3. Sure.

Comment (mmazur) [reply lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py]: 1. Ok, so your point is that it was intended to work with VMIs, it just had a bunch of issues. In that case, I'll extend the waiting to work with vmis as well.
2. Hehe, I'm a dumb dumb and didn't check the underlying code. KubernetesRawModule.perform_action() (which gets called within execute_crud()) is actually wait–aware and will wait for resource creation/deletion. With creation we need additional logic on top, since we're not interested in just the k8s object being created, but the VMI actually running (or being stopped in case of VM with status: stopped). However in the case of VMI deletion, one perform_action returns, that actually means the object was actually deleted and there's nothing else to wait for.

Bottom line: I'll add VMI running waiting and that will mean both ephemeral and non–ephemeral vms will have rock–solid wait support.

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py (hunk 2 above)
Comment (mmazur) [reply lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py:357]: It's not that complex, the whole logic is ~10 lines of additional code. And the benefit is that in the 99% of cases where you just intend to start/stop a vm (like [1] where there are multiple other  examples)  you get a human–friendly error that clearly hints you're probably have your name wrong. Were it 100 lines of code, I wouldn't be so sure, but in this case the UX benefit is clear and the additional code isn't much.

[1] https://github.com/kubevirt/ansible-kubevirt-modules/blob/master/tests/playbooks/e2e.yaml#L41

Comment (machacekondra) [reply lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py:357]: I am fine. It's just that it's `if` with too many and\or's are just too much for me :) I just preffer straighforward code with no complex if's, if they are not really needed.

**Comment by mmazur** [comment]
And done. As far as I can tell everything is working correctly. I'll add a PR extending our travis playbooks to make sure all of this always gets solidly tested. Other than that, once the ansible CI finishes, you can shipit.

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py [hunk 3]
```diff
@@ -247,92 +252,118 @@ def argspec(self):
...
+        for event in self._kind_resource.watch(namespace=self.namespace, timeout=self.params.get('wait_timeout')):
+            entity = event['object']
+            if entity.metadata.name != self.name:
+                continue
+            status = entity.get('status', {})
+            phase = status.get('phase', None)
+            if phase == 'Running':
+                return entity
+
+        self.fail("Error waiting for virtual machine. Maybe try a higher wait_timeout value?")
```
Comment (pkliczewski) [inline lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py]: What about `Timeout occurred while waiting on virtual machine to start`?

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py [hunk 4]
```diff
@@ -247,92 +252,118 @@ def argspec(self):
...
+            entity = event['object']
+            if entity.metadata.name != self.name:
+                continue
+            status = entity.get('status', {})
+            created = status.get('created', False)
+            ready = status.get('ready', False)
+            if (created, ready) == (want_created, want_ready):
+                return entity
+
+        self.fail("Error waiting for virtual machine. Maybe try a higher wait_timeout value?")
```
Comment (pkliczewski) [inline lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py]: What about `Timeout occurred while waiting on virtual machine to get to %s state`?

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py [hunk 5]
```diff
@@ -247,92 +252,118 @@ def argspec(self):
...
+    @staticmethod
+    def fix_serialization(obj):
+        if obj and hasattr(obj, 'to_dict'):
+            return obj.to_dict()
+        return obj
+
+    def _wait_for_vmi_running(self):
+        for event in self._kind_resource.watch(namespace=self.namespace, timeout=self.params.get('wait_timeout')):
+            entity = event['object']
+            if entity.metadata.name != self.name:
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py:267]: please reuse the code, it's very same

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py [hunk 6]
```diff
@@ -247,92 +252,118 @@ def argspec(self):
...
+            result = self.execute_crud(kind, definition)
+            changed = result['changed']
+            k8s_obj = result['result']
+            method = result['method']
+
+        if ephemeral and state == 'present' and self.params.get('wait') and not self.check_mode:
+            # Waiting for state==absent is handled inside execute_crud()
+            k8s_obj = self._wait_for_vmi_running()
+
+        if not ephemeral and state in ['running', 'stopped'] and not self.check_mode:
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py]: I think it make sense to support running state for ephmeral VM, as well as stopped and absent. which would be same. The code will be more straighforward then without not needed ifs

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py [hunk 7]
```diff
@@ -247,92 +252,118 @@ def argspec(self):
...
+                return entity
+
+        self.fail("Error waiting for virtual machine. Maybe try a higher wait_timeout value?")
+
+    def manage_vm_state(self, new_state, already_changed):
+        new_running = True if new_state == 'running' else False
+        changed = False
+        k8s_obj = {}
+
+        if not already_changed:
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py:300]: So if I update some parameter+ change state to running, it won't start, IIUC

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py [hunk 8]
```diff
@@ -29,10 +29,11 @@
...
             - Set the virtual machine to either I(present), I(absent), I(running) or I(stopped).
-            - "I(present) - Create or update virtual machine."
-            - "I(absent) - Removes virtual machine."
-            - "I(running) - Create or update virtual machine and run it."
-            - "I(stopped) - Stops the virtual machine."
+            - "Ephemeral virtual machines do not support I(running)/I(stopped) states."
+            - "I(present) - Create or update a virtual machine."
+            - "I(absent) - Remove a virtual machine."
+            - "I(running) - Create or update a non-ephemeral virtual machine and run it."
+            - "I(stopped) - Stop a non-ephemeral virtual machine."
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py]: Not sure if there is use case for it , but there may be use case that user want to ensure VM is created and stopped, I would still prefer to support it, if it used to work. And again we get rid of many complex if's

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py (hunk 7 above)
Comment (mmazur) [reply lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py:300]: If you change state to running, then `execute_crud()` will patch the object and change state to running, which will make it start the VMI. At the same time `already_changed` will be set to `True`, which will skip patching the object for a second time, because why patch the object for a second time if we don't have to.

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py (hunk 6 above)
Comment (mmazur) [reply lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py]: @pkliczewski we need a tie breaker.

In my code ephemeral vms (aka VMIs) support only states _present_ (VMI exists) and _absent_ (VMI doesn't exist), while normal/non–ephemeral ones (VM objects) support also _running_ and _stopped_. My reasoning is that if ephemeral vms also supported running/stopped, that would mean adding two aliases: running==present (which makes sense for a VMI), but also stopped==absent.

And the latter I really don't like. Without those aliases, if you want to stop an ephemeral vm, you're forced to do `state: absent`, which makes it clear you're actually deleting it. With the aliases added an operator might attempt `state: stopped` and be surprised that the vm got deleted and not stopped. (My imaginary operator is someone accustomed to traditional virt systems, for whom such behavior is non–obvious and surprising.)

We need someone to call on this. Three options, choose wisely:
1. ephemerals support only present/absent; no aliases for running/stopped
2. ephemerals support present/running/absent, with running==present; no _stopped_ to avoid the nasty surprise
3. ephemerals support present/running/absent/stopped, with running==present and stopped==absent

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py (hunk 3 above)
Comment (mmazur) [reply lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py]: done

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py (hunk 4 above)
Comment (mmazur) [reply lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py]: done

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py (hunk 5 above)
Comment (mmazur) [reply lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py:267]: It looks like it on first glance, but I'd actually end up with more lines of (less legible) code if I tried that.

**Comment by mmazur** [comment]
Ok, I've rebased the patch on top of the template changes and added a new function `construct_definition()` as `execute_module()` has gotten too long. Now waiting on Piotr to make a call.

**Review by machacekondra — COMMENTED** [review]
shipit

File: lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py (hunk 6 above)
Comment (pkliczewski) [reply lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py]: @mmazur I think that Ondra's suggestion is correct. We need to be consistent how we present vm state to the user. At some point users will be confused if they not understand ephemeral vs stateful vm. For ephemeral vms stopping means delete and it is fine.

**Comment by mmazur** [comment]
Right, turns out there were still a few corner cases I've missed. Should be ready in a few hours.

**Comment by mmazur** [comment]
And done. I'll go write a playbook for our repo to make sure all the corner cases I currently have in my head are actually properly tested and verified working

**Comment by pkliczewski** [comment]
:+1:

**Comment by mmazur** [comment]
Ok, so https://github.com/kubevirt/ansible-kubevirt-modules/pull/218 contains a comprehensive list of all the stuff this code does and tests for whether it does them correctly. That whole playbook passes on my system. So as far as I can tell this code is correct. Please add shipits and let's wait for someone to merge.

**Comment by pkliczewski** [comment]
+1

**Comment by gundalow** [comment]
@mmazur Can you please rebase looks like one of the PRs I've just merged has caused a conflict
After rebase we will need `shipit` adding again to get `label/shipit` which is what triggers someone to merge, thanks

**Comment by mmazur** [comment]
@machacekondra @pkliczewski had to rebase, please +1 again

**Comment by pkliczewski** [comment]
+1

**Comment by gundalow** [comment]
Thanks for the quick rebase.

Merged into `devel` for release in Ansible 2.8
